Fixes for Windows.

maleadt opened the pull request:
Fixes #4.

maleadt committed:
Fixes for Windows.

maleadt committed:
Proper detection of MSVC

musm commented on src/discovery.jl:
```diff
@@ -269,13 +278,26 @@ function find_toolchain(toolkit_path, toolkit_version=find_toolkit_version(toolk
+        # NOTE: we could try and detect cl.exe from the env vars below,
+        #       but Microsoft seems to change the location often.
+        #       so we just require the user to modify his PATH
+        # envvars = collect(filter(key->ismatch(r"^VS\d+.+TOOLS$", key), keys(ENV)))
```
Those should still be fine for now, it's just that Visual studio 2017 changed things.

maleadt replied:
let's just rely on `vswhere` for all versions?

musm replied:
`vswhere` is only for 2017 and newer however, it doesn't seem to find my  VS 2015 install. 
Indeed looking at the description at  https://github.com/Microsoft/vswhere
"Locate Visual Studio 2017 and newer installations"

musm replied:
Perhaps it would also just be better to use `%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe` instead of downloading the exe, since it's included by default in versions starting from 15.2 and newer

maleadt replied:
Yeah, I thought it detected older versions hence the download. What a mess.

musm replied:
I can confirm  this PR works with cuda 9 and vs studio 2017

denizyuret commented:
I could not get this to work with CUDA 9 and VS 2017.  Here is the version compatibility info from CUDA (CUDA/v9.0/include/crt/host_config.h):
```
#if _MSC_VER < 1600 || _MSC_VER > 1911
#error -- unsupported Microsoft Visual Studio version! Only the versions 2012, 2013, 2015 and 2017 are supported!
```
And VS 2017 came with _MSC_VER > 1911 (presumably they just released it after cuda 9). So I downgraded to VS 2015 which has version 1900. `vswhere.exe` cannot seem to find VS 2015.

maleadt commented:
Ah, so version checking will be messier. Probably have to check the value of `MSC_VER` by calling into `cl.exe`. I checked the CUDA versions I have around:

```julia
# Source: CUDA/include/host_config.h or include/crt/host_config.h on 9.0+
#
# NOTE: plain integers, because from MSVC 14.1 on the number increases monolithically
#       https://blogs.msdn.microsoft.com/vcblog/2016/10/05/visual-c-compiler-version/
const cuda_msvc_db = Dict(
    v"5.5"     => 1500:1700,  # (_MSC_VER < 1500 || _MSC_VER > 1700) && #error
    v"6.0"     => 1500:1700,  # (_MSC_VER < 1500 || _MSC_VER > 1700) && #error
    v"6.5"     => 1600:1800,  # (_MSC_VER < 1600 || _MSC_VER > 1800) && #error
    v"7.0"     => 1600:1800,  # (_MSC_VER < 1600 || _MSC_VER > 1800) && #error
    v"7.5"     => 1600:1800,  # (_MSC_VER < 1600 || _MSC_VER > 1800) && #error
    v"8.0"     => 1600:1900,  # (_MSC_VER < 1600 || _MSC_VER > 1900) && #error
    v"9.0"     => 1600:1910   # (_MSC_VER < 1600 || _MSC_VER > 1910) && #error
    v"9.0.176" => 1600:1911   # (_MSC_VER < 1600 || _MSC_VER > 1911) && #error
)
```

Also, feel free to contribute to this branch/PR, I only very occasionally have access to a Windows set-up.

maleadt committed:
Add fixes from #11.

maleadt committed:
Fix the MSVC version compat db, even though it is unused.

maleadt committed:
Add some tests.

maleadt committed:
Add cuDNN version database and name.

maleadt committed:
Fixes.

maleadt commented:
@denizyuret I imported your fixes VS detection logic from #11 to this branch.
I'm not happy with it, but (not being a Windows user) I don't care enough to improve it.
Capturing STDERR is possible with `pipeline` and `Pipe`.

As I already moved the special-casing for adding version digits to DLL names out of `find_library`, I also changed the logic you added there for cuDNN. It should be possible to do `find_library(CUDAapi.libcudnn)`, where `CUDAapi.libcudnn` now is a library of candidate names.

Does that still work? Feel free to push changes to this branch!

denizyuret commented:
@maleadt what is on the todo list to merge this with master and register a new version?

maleadt commented:
Nothing, really. I had hoped the MSVC detection would be a little more robust, ie. matching any available version against the CUDA database I added in `compatibility.jl` (because without it, we're bound to get bug reports about compilation failures), but for now it's a strict improvement. Also, with packages like Knet hopefully relying on CUDAnative's codegen we won't need to use this part of CUDAapi.jl.

Maybe @musm could you verify whether this works on your set-up(s)? If so, feel free to merge. I'll do a double-check on my systems and tag a new release shortly after.

musm commented:
@maleadt Is this  supposed  find  `cudart32_*` versions by any chance? Seems like all the libcudart are for `cuda64*`

BTW there are some issues with this PR I'm trying to fix

maleadt commented:
> @maleadt Is this supposed find cudart32_* versions by any chance? Seems like all the libcudart are for cuda64*

It should, on a 32-bit system. This should populate an array with the possible names: https://github.com/JuliaGPU/CUDAapi.jl/blob/f8602794a68c0e71f50692e16c446f3bd74456da/src/discovery.jl#L11-L16

Should be invoked through `find_library(CUDAapi.libcudart)`, as in: https://github.com/JuliaGPU/CUDAapi.jl/blob/f8602794a68c0e71f50692e16c446f3bd74456da/src/discovery.jl#L178

musm commented:
Hi @maleadt  I needed the following modifications for the search part (no changes yet to the finding of VS studio)

```patch
diff --git a/src/discovery.jl b/src/discovery.jl
index d497fab..55abd93 100644
--- a/src/discovery.jl
+++ b/src/discovery.jl
@@ -40,17 +40,22 @@ function find_library(names::Vector{String}, prefixes::Vector{String}=String[])
     end
 
     # figure out locations
-    locations = []
+    locations = String[]
     for prefix in prefixes
         push!(locations, prefix)
-        push!(locations, joinpath(prefix, "lib"))
-        if Sys.WORD_SIZE == 64
-            push!(locations, joinpath(prefix, "lib64"))
+        if !Compat.Sys.iswindows()
+            push!(locations, joinpath(prefix, "lib"))
+            if Sys.WORD_SIZE == 64
+                push!(locations, joinpath(prefix, libfoldername64))
+            end
+        else
+            push!(locations, joinpath(prefix, "bin"))
         end
     end
 
     @trace("Checking for $names in $locations")
-    name = Libdl.find_library(names, locations)
+    # we do reverse since we want to search the latest version first!
+    name = Libdl.find_library(reverse!(names), locations)
     if isempty(name)
         error("Could not find any of $names")
     end
@@ -76,7 +81,7 @@ function find_binary(name, prefixes::Vector{String}=String[])
     end
 
     # figure out locations
-    locations = []
+    locations = String[]
     for prefix in prefixes
         push!(locations, prefix)
         push!(locations, joinpath(prefix, "bin"))
@@ -147,9 +152,22 @@ end
 
 function find_toolkit()
     # figure out locations
-    dirs = ["/usr/lib/nvidia-cuda-toolkit",
-            "/usr/local/cuda",
-            "/opt/cuda"]
+    if !Compat.Sys.iswindows()
+        dirs = ["/usr/lib/nvidia-cuda-toolkit",
+                "/usr/local/cuda",
+                "/opt/cuda"]
+    else
+        # CUDA versions are installed in seperate directories under the following root CUDA folder by default
+        locations = [joinpath(get(ENV, "ProgramFiles", ""), "NVIDIA GPU Computing Toolkit\\CUDA"),
+                      joinpath(get(ENV, "ProgramFiles(x86)", ""), "NVIDIA GPU Computing Toolkit\\CUDA")]
+        filter!(isdir, locations)
+        dirs = String[]
+        for dir in locations
+            # add all CUDA toolkit versions under the root CUDA folder
+            push!(dirs, map(x -> joinpath(dir, x),  readdir(dir))...)
+        end
+        sort!(dirs, rev=true) # we want to search starting from the newest CUDA version
+    end
     ## look for environment variables (taking priority over default values)
     envvars = ["CUDA_PATH", "CUDA_HOME", "CUDA_ROOT"]
     envvars_set = filter(var -> haskey(ENV, var), envvars)
@@ -163,7 +181,7 @@ function find_toolkit()
     end
     ## look for the compiler binary (in the case PATH points to the installation)
     try
-        nvcc_path = find_binary(nvcc)
+        nvcc_path = find_binary(nvcc, dirs)
         dir = dirname(nvcc_path)
         if ismatch(r"^bin(32|64)?$", basename(dir))
             dir = dirname(dir)
@@ -175,9 +193,9 @@ function find_toolkit()
     end
     ## look for the runtime library (in the case LD_LIBRARY_PATH points to the installation)
     try
-        libcudart_path = find_library(libcudart)
+        libcudart_path = find_library(libcudart, dirs)
         dir = dirname(libcudart_path)
-        if ismatch(r"^lib(32|64)?$", basename(dir))
+        if ismatch(r"^(lib|bin)(32|64)?$", basename(dir))
             dir = dirname(dir)
         end
         @trace("Considering CUDA toolkit at $dir based on libcudart at $libcudart_path")


```

musm commented:
Here is the updated patch.

This should cover all cases, hopefully

```patch
diff --git a/src/discovery.jl b/src/discovery.jl
index d497fab..df8756b 100644
--- a/src/discovery.jl
+++ b/src/discovery.jl
@@ -40,17 +40,22 @@ function find_library(names::Vector{String}, prefixes::Vector{String}=String[])
     end
 
     # figure out locations
-    locations = []
+    locations = String[]
     for prefix in prefixes
         push!(locations, prefix)
-        push!(locations, joinpath(prefix, "lib"))
-        if Sys.WORD_SIZE == 64
-            push!(locations, joinpath(prefix, "lib64"))
+        if !Compat.Sys.iswindows()
+            push!(locations, joinpath(prefix, "lib"))
+            if Sys.WORD_SIZE == 64
+                push!(locations, joinpath(prefix, libfoldername64))
+            end
+        else
+            push!(locations, joinpath(prefix, "bin"))
         end
     end
 
     @trace("Checking for $names in $locations")
-    name = Libdl.find_library(names, locations)
+    # we do reverse since we want to search the latest version first!
+    name = Libdl.find_library(reverse!(names), locations)
     if isempty(name)
         error("Could not find any of $names")
     end
@@ -76,7 +81,7 @@ function find_binary(name, prefixes::Vector{String}=String[])
     end
 
     # figure out locations
-    locations = []
+    locations = String[]
     for prefix in prefixes
         push!(locations, prefix)
         push!(locations, joinpath(prefix, "bin"))
@@ -147,9 +152,22 @@ end
 
 function find_toolkit()
     # figure out locations
-    dirs = ["/usr/lib/nvidia-cuda-toolkit",
-            "/usr/local/cuda",
-            "/opt/cuda"]
+    if Compat.Sys.iswindows()
+        # CUDA versions are installed in seperate directories under the following root CUDA folder
+        searchdirs = [joinpath(get(ENV, "ProgramFiles", ""), "NVIDIA GPU Computing Toolkit\\CUDA"),
+                      joinpath(get(ENV, "ProgramFiles(x86)", ""), "NVIDIA GPU Computing Toolkit\\CUDA")]
+        filter!(isdir, searchdirs)
+        dirs = String[]
+        for dir in searchdirs
+            # add all CUDA toolkit versions under the root CUDA folder
+            push!(dirs, map(x -> joinpath(dir, x),  readdir(dir))...)
+        end
+        sort!(dirs, rev=true) # we want to search starting from the newest CUDA version
+    else
+        dirs = ["/usr/lib/nvidia-cuda-toolkit",
+                "/usr/local/cuda",
+                "/opt/cuda"]
+    end
     ## look for environment variables (taking priority over default values)
     envvars = ["CUDA_PATH", "CUDA_HOME", "CUDA_ROOT"]
     envvars_set = filter(var -> haskey(ENV, var), envvars)
@@ -163,7 +181,7 @@ function find_toolkit()
     end
     ## look for the compiler binary (in the case PATH points to the installation)
     try
-        nvcc_path = find_binary(nvcc)
+        nvcc_path = find_binary(nvcc, dirs)
         dir = dirname(nvcc_path)
         if ismatch(r"^bin(32|64)?$", basename(dir))
             dir = dirname(dir)
@@ -175,9 +193,9 @@ function find_toolkit()
     end
     ## look for the runtime library (in the case LD_LIBRARY_PATH points to the installation)
     try
-        libcudart_path = find_library(libcudart)
+        libcudart_path = find_library(libcudart, dirs)
         dir = dirname(libcudart_path)
-        if ismatch(r"^lib(32|64)?$", basename(dir))
+        if ismatch(r"^(lib|bin)(32|64)?$", basename(dir))
             dir = dirname(dir)
         end
         @trace("Considering CUDA toolkit at $dir based on libcudart at $libcudart_path")
@@ -194,9 +212,9 @@ function find_toolkit()
     end
 
     # select
-    dir = first(dirs)
-    @debug("Using CUDA toolkit at $dir")
-    return dir
+    cudadir = first(dirs)
+    @debug("Using CUDA toolkit at $cudadir")
+    return cudadir
 end
 
 # figure out the CUDA toolkit version (by looking at the `nvcc --version` output)
@@ -293,18 +311,62 @@ function find_host_compiler(toolkit_version=nothing)
         sort!(gcc_possibilities; rev=true, lt=(a, b) -> a[2]<b[2])
         host_compiler, host_version = gcc_possibilities[1]
     elseif Compat.Sys.iswindows()
-        # Windows: just use cl.exe
-        # TODO: discovery of all compilers, and version matching against cuda_msvc_db
-        vc_versions = ["VS140COMNTOOLS", "VS120COMNTOOLS", "VS110COMNTOOLS", "VS100COMNTOOLS"]
-        !any(x -> haskey(ENV, x), vc_versions) && error("Compatible Visual Studio installation cannot be found; Visual Studio 2015, 2013, 2012, or 2010 is required.")
-        vs_cmd_tools_dir = ENV[vc_versions[first(find(x -> haskey(ENV, x), vc_versions))]]
-        msvc_path = joinpath(dirname(dirname(dirname(vs_cmd_tools_dir))), "VC", "bin", Sys.WORD_SIZE == 64 ? "amd64" : "", "cl.exe")
-        tmpfile = tempname() # cl.exe writes version info to stderr, this is the only way I know of capturing it
-        readstring(pipeline(`$msvc_path`,stderr=tmpfile))
-        msvc_ver_str = match(r"Version\s+(\d+(\.\d+)?(\.\d+)?)"i, readstring(tmpfile))[1]
-        msvc_ver = VersionNumber(msvc_ver_str)
-
-        host_compiler, host_version = msvc_path, msvc_ver
+        # Windows: just use cl.exe and search for MSVS in default locations
+        try
+            global msvs_paths
+            msvs_paths = String[]
+            arch = Sys.WORD_SIZE == 64 ? "amd64" : "x86"
+
+            # locate VS2017
+            vswhere = joinpath(get(ENV, "ProgramFiles(x86)", ""), "Microsoft Visual Studio", "Installer", "vswhere.exe")
+            if isfile(vswhere)
+                msvs_cmd_tools_dir = chomp(read(`$vswhere -latest -property installationPath`, String))
+                vs_prompt = joinpath(msvs_cmd_tools_dir, "VC", "Auxiliary", "Build", "vcvarsall.bat")
+                tmpfile = tempname()
+                run(pipeline(`"$vs_prompt" $arch \& where cl.exe`, tmpfile))
+                msvs_path = readlines(tmpfile)[end]
+                push!(msvs_paths, msvs_path)
+            end
+
+            # locate VS2012 to 2014
+            vc_versions_100_140 = ["VS140COMNTOOLS", "VS120COMNTOOLS", "VS110COMNTOOLS", "VS100COMNTOOLS"]
+            for inds in find(x -> haskey(ENV, x), vc_versions_100_140)
+                path = ENV[vc_versions_100_140[inds]]
+                msvs_path = joinpath(dirname(dirname(dirname(path))), "VC", "bin", arch, "cl.exe")
+                push!(msvs_paths, msvs_path)
+            end
+        catch ex
+            error("Compatible Visual Studio installation cannot be found; Visual Studio 2017, 2015, 2013, 2012, or 2010 is required.")
+        end
+
+        msvs_list = Dict{VersionNumber,String}()
+        for path in msvs_paths
+            tmpfile = tempname() # cl.exe writes version info to stderr, this is the only way I know of capturing it
+            read(pipeline(`$path`, stderr=tmpfile), String)
+            ver_str = match(r"Version\s+(\d+(\.\d+)?(\.\d+)?)"i, read(tmpfile, String))[1]
+            ver = VersionNumber(ver_str)
+            msvs_list[ver] = path
+        end
+
+        if toolkit_version !== nothing
+            found_compatible_compiler = false
+            for ver in sort(collect(keys(msvs_list)), rev=true) # search the higest version first
+                if parse(Int, string(ver.major,ver.minor)) in cuda_msvc_db[toolkit_version]
+                    host_compiler, host_version = msvs_list[ver], ver
+                    found_compatible_compiler = true
+                    break
+                end
+            end
+            if !found_compatible_compiler
+                error("Visual Studio C++ compiler $msvs_ver is not compatible with CUDA $toolkit_version")
+            end
+            return host_compiler, host_version
+        else
+            # take the first found host, which will be the higest version found
+            host_pair = first(sort(collect(msvs_list), by=x->x[1], rev=true))
+            host_compiler, host_version = last(host_pair), first(host_pair)
+            return host_compiler, host_version
+        end
     elseif Compat.Sys.isapple()
         # GCC is no longer supported on MacOS so let's just use clang
         # TODO: discovery of all compilers, and version matching against the toolkit

```

musm commented:
feel free to push the changes and modify some of the logic if you can make it simpler

musm committed:
Further improve Windows detection.

maleadt commented:
@musm thanks! Did some minimal editing, should be all set.

musm commented on src/discovery.jl:
```diff
@@ -256,36 +295,89 @@ function find_toolchain(toolkit_path, toolkit_version=find_toolkit_version(toolk
+                msvs_cmd_tools_dir = chomp(read(`$vswhere -latest -property installationPath`, String))
+                vs_prompt = joinpath(msvs_cmd_tools_dir, "VC", "Auxiliary", "Build", "vcvarsall.bat")
+                tmpfile = tempname()
+                run(pipeline(`"$vs_prompt" $arch \& where cl.exe`, tmpfile))
```
we can remove the quotes around `"$vs_prompt"`

musm committed:
Remove unneccesary quotes.

musm commented:
Great should we then merge this, and make any updates according to further issues encountered?
